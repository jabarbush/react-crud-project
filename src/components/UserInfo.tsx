import React, { useState } from 'react';
import UserInfoProps from '../interfaces/UserInfoProps';
import { ReactComponent as Trash } from '../assets/trash.svg';
import { ReactComponent as Edit } from '../assets/edit.svg';
import { ReactComponent as FormClose } from '../assets/form-close.svg';
import DeleteUser from './DeleteUser';

const UserInfo: React.FC<UserInfoProps> = ({ userList, selectedUser, onCloseUserInfo }) => {
    const [showDeleteUser, setShowDeleteUser] = useState(false);

    const handleOpenDeleteUser = () => {
        setShowDeleteUser(true);
      };
    
      const handleCloseDeleteUser = () => {
        setShowDeleteUser(false);
      };

  return (
    <div className={`info-tab ${selectedUser ? 'active' : ''}`}>
        <div className='data'>
            <div className='user-name'>{selectedUser?.name}</div>
            <p className='user-info'>
                DOB:&nbsp;
                <div className='user-info-value'>
                    {selectedUser?.dob}
                </div>
            </p>
            <p className='user-info'>
                Phone:&nbsp;
                <div className='user-info-value'>
                    {selectedUser?.phone}
                </div>
            </p>
            <p className='user-info'>
                Email:&nbsp;
                <div className='user-info-value'>
                    {selectedUser?.email}
                </div>
            </p>
            <div className='icons'>
                <Trash className='icon' onClick={handleOpenDeleteUser} />
                <Edit className='icon'/>
                <FormClose onClick={onCloseUserInfo} className='icon' />
            </div>
        </div>
        <hr/>
        <div className='signature-form'>
            Add a custom signature
            <label>Signature PIN (6 Digits)</label>
            <br/>
            <input type="text" placeholder='Signature PIN'/>
            <br/>
            <label>Signature Font Style</label>
            <br/>
            <input type="text" placeholder='Select font...'/>
            <br/>
            <label>Signature Preview</label>
            <br/>
            <div className='signature-preview'>
            </div>
        </div>
        {showDeleteUser && 
        <DeleteUser userList={userList} selectedUser={selectedUser} onClose={handleCloseDeleteUser} onDeleteUser={onCloseUserInfo}/>}
    </div>
  );
};

export default UserInfo;
