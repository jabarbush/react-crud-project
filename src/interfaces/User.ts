interface User {
    id: number;
    name: string;
    dob: string;
    phone: string;
    email: string;
    signature?: {
        pin: string;
        fontStyle: string;
    } | null;
  }
  
  export default User;
  