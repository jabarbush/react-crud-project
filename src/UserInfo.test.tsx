import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import UserInfo from './components/UserInfo';
import User from './interfaces/User';

beforeEach(() => jest.resetAllMocks())
describe('UserInfo component', () => {
  const userList: User[] = [];
  const selectedUser = {
    id: 4,
    name: 'jake',
    dob: '03/03/2003',
    phone: '4564564564',
    email: 'jake4@gmail.com',
    signature: {
      pin: '123123',
      fontStyle: 'Mr Dafoe',
    },
  };
  const selectedUser_null_signature = {
    id: 4,
    name: 'jake',
    dob: '03/03/2003',
    phone: '4564564564',
    email: 'jake4@gmail.com',
    signature: null,
  };
  const onCloseUserInfo = jest.fn();
  const setUserList = jest.fn();

  test('renders user information correctly', () => {
    render(
      <UserInfo
        userList={userList}
        selectedUser={selectedUser}
        onCloseUserInfo={onCloseUserInfo}
        setUserList={setUserList}
      />
    );

    const userNameElement = screen.getByText(selectedUser.name);
    const dobElement = screen.getByText(`${selectedUser.dob}`);
    const phoneElement = screen.getByText(`${selectedUser.phone}`);
    const emailElement = screen.getByText(`${selectedUser.email}`);
    const trashIconElement = screen.getByTestId('delete-user-icon');
    const editIconElement = screen.getByTestId('edit-user-icon');

    expect(userNameElement).toBeInTheDocument();
    expect(dobElement).toBeInTheDocument();
    expect(phoneElement).toBeInTheDocument();
    expect(emailElement).toBeInTheDocument();
    expect(trashIconElement).toBeInTheDocument();
    expect(editIconElement).toBeInTheDocument();
    expect(screen.getByTestId('userInfo-component')).toMatchSnapshot();
  });

  test('opens delete user dialog on trash icon click', () => {
    render(
      <UserInfo
        userList={userList}
        selectedUser={selectedUser}
        onCloseUserInfo={onCloseUserInfo}
        setUserList={setUserList}
      />
    );

    const trashIconElement = screen.getByTestId('delete-user-icon');
    fireEvent.click(trashIconElement);

    const deleteDialogElement = screen.getByText('Confirm Delete');
    expect(deleteDialogElement).toBeInTheDocument();
  });

  test('opens edit user info form on edit icon click', () => {
    render(
      <UserInfo
        userList={userList}
        selectedUser={selectedUser}
        onCloseUserInfo={onCloseUserInfo}
        setUserList={setUserList}
      />
    );

    const editIconElement = screen.getByTestId('edit-user-icon');
    fireEvent.click(editIconElement);

    const editFormElement = screen.getByText('Edit User Information');
    expect(editFormElement).toBeInTheDocument();
  });

  test('adds a signature if one is not present', () => {
    render(
      <UserInfo
        userList={userList}
        selectedUser={selectedUser_null_signature}
        onCloseUserInfo={onCloseUserInfo}
        setUserList={setUserList}
      />
    );

    const addSignatureButton = screen.getByText('Add Signature');
    fireEvent.click(addSignatureButton);

    
    expect(selectedUser_null_signature).not.toBeNull;
  });
});
