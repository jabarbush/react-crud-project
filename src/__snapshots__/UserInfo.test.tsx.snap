// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInfo component renders user information correctly 1`] = `
<div
  class="info-tab"
  data-testid="userInfo-component"
>
  <div>
    <div
      class="data"
    >
      <div
        class="user-name"
      >
        jake
      </div>
      <div
        class="user-info"
        data-testid="date-of-birth"
      >
        DOB: 
        <div
          class="user-info-value"
        >
          03/03/2003
        </div>
      </div>
      <div
        class="user-info"
        data-testid="phone"
      >
        Phone: 
        <div
          class="user-info-value"
        >
          4564564564
        </div>
      </div>
      <div
        class="user-info"
        data-testid="email"
      >
        Email: 
        <div
          class="user-info-value"
        >
          jake4@gmail.com
        </div>
      </div>
      <div
        class="icons"
      >
        <svg
          class="icon"
          data-testid="delete-user-icon"
        >
          trash.svg
        </svg>
        <svg
          class="icon"
          data-testid="edit-user-icon"
        >
          edit.svg
        </svg>
        <svg
          class="icon-big-x"
        >
          form-close.svg
        </svg>
      </div>
    </div>
    <hr />
    <div>
      <div
        class="signature-status-section"
      >
        <h3
          class="signature-status-title"
        >
          Signature status:
        </h3>
        <div
          class="signature-status-info green-info"
        >
          Has Custom Signature
        </div>
        <button
          class="signature-view-edit-btn"
        >
          View Signature
        </button>
      </div>
      <hr />
    </div>
  </div>
</div>
`;
